_summaries/beltagy2020longformer_2.md

```diff
@@ -0,0 +1,146 @@
+
+    <p align="center">
+        <img src="beltagy2020longformer/2a.png" width="15%" height="15%">
+    </p>
+
+    **Sliding Window**: Instead of having a query token attend to every query token, the query token can attend to a fixed window size $$w$$ of tokens near itself, which computes some attention in $$\mathcal{O}(n w d)$$. Each token attends to $$\frac{1}{2}w$$ tokens on each side, like the following multiplication pattern:
+
+    <p align="center">
+        <img src="beltagy2020longformer/2b.png" width="15%" height="15%">
+    </p>
+
+    Using multiple stacked layers of such windowed attention can result in a large receptive field. Thus, for a transformer with $$l$$ layers, the receptive field at the top layer would be $$l \times w$$.
+
+    **Dilated Sliding Window**: Sliding windows can be "dilated" to increase the receptive field without increasing computation like the following multiplication pattern:
+
+    <p align="center">
+        <img src="beltagy2020longformer/2c.png" width="15%" height="15%">
+    </p>
+
+    If $$d$$ is the dilation size, then the receptive field is $$l \times d \times w$$. In a multi-headed attention setting, different heads can have different dilation sizes for heads to focus on different sizes of contexts.
+
+    **Global Attention**: The window and dilated attention may not be flexible enough to learn task-specific representations. For example, in the case of a `[CLS]` token, it wouldn't make sense for it to attend to only the first few tokens of input. Thus, some pre-selected input locations will still compute "global attention". That is, a token with a global attention attends to all tokens across the sequence, and all tokens in the sequence attend to it. This looks like the following multiplication pattern:
+
+    <p align="center">
+        <img src="beltagy2020longformer/2d.png" width="15%" height="15%">
+    </p>
+
+    In the case of a QA task, global attention is provided on all question tokens. Because the number of tokens with global attention is small relative to sequence length, the combined local and global attention is still bounded by $$\mathcal{O}(n)$$.
+
+    In implementation, there's a linear projection for sliding window attention and a separate linear projection for global attention per token. This method avoids the global and sliding window attentions from convoluting each other.
+
+    **Implementation Detail**: Modern PyTorch/Tensorflow libraries do not support the form of banded matrix multiplication required for the sliding window attention, so the authors implemented 3 ways of doing it: `loop`, `chunks`, and `cuda`. `loop` is used for testing, `chunks` is used for pre-training/fine-tuning, `cuda` is a highly optimized custom CUDA kernel that is used for the language modeling experiments.
```

Comment:
Might be worthwhile to mention that different Q,K,V projections are used for computing local (sliding or dilated) versus global attention